File: src/lib/cli/commands/governance/delivery/control-recovery.mjs
import { ArcaneError } from '../../../../contracts/arcane/errors.mjs';

const RECOVERY_SCHEMA = 'arcane.control-recovery.v1';

function requireObject(value, name) {
  if (!value || typeof value !== 'object' || Array.isArray(value)) {
    throw new ArcaneError('ARC_SCHEMA_INVALID', `${name} must be an object`, { name });
  }
  return value;
}

function requireId(value, name) {
  if (typeof value !== 'string' || value.length === 0) {
    throw new ArcaneError('ARC_SCHEMA_INVALID', `${name} must be a non-empty string`, { name });
  }
  return value;
}

/**
 * Execute one narrowly authorized control-state recovery.  Callers retain
 * persistence ownership: this function returns a replacement or quarantine
 * disposition; it never writes a control store itself.
 */
export function recoverControlState({ controlId, state, authorization, authenticate, repair, verify }) {
  requireId(controlId, 'controlId');
  requireObject(authorization, 'authorization');
  if (authorization.controlId !== controlId || authorization.operation !== 'RECOVER_CONTROL_STATE') {
    return Object.freeze({ schema: RECOVERY_SCHEMA, allowed: false, code: 'ARC_RECOVERY_SCOPE_DENIED', controlId });
  }
  if (typeof authenticate !== 'function' || authenticate(authorization) !== true) {
    return Object.freeze({ schema: RECOVERY_SCHEMA, allowed: false, code: 'ARC_RECOVERY_AUTH_REQUIRED', controlId });
  }
  if (typeof repair !== 'function' || typeof verify !== 'function') {
    throw new ArcaneError('ARC_SCHEMA_INVALID', 'recoverControlState requires repair and verify functions', {});
  }

  let recovered;
  try {
    recovered = repair(state, Object.freeze({ controlId, authorization }));
  } catch (error) {
    return Object.freeze({
      schema: RECOVERY_SCHEMA, allowed: false, code: 'ARC_RECOVERY_QUARANTINED', controlId,
      quarantine: Object.freeze({ originalState: state, reasonCode: error?.code ?? 'ARC_RECOVERY_REPAIR_FAILED' }),
    });
  }
  const verification = verify(recovered, Object.freeze({ controlId, authorization }));
  if (verification !== true) {
    return Object.freeze({
      schema: RECOVERY_SCHEMA, allowed: false, code: 'ARC_RECOVERY_VERIFY_FAILED', controlId,
      quarantine: Object.freeze({ originalState: state, recoveredState: recovered, reasonCode: 'ARC_RECOVERY_VERIFY_FAILED' }),
    });
  }
  return Object.freeze({
    schema: RECOVERY_SCHEMA, allowed: true, code: 'ARC_RECOVERY_RESUMED', controlId,
    replacementState: recovered, preservedState: state, verification: 'PASS',
  });
}

File: engine/crates/legion-arcane/src/control_recovery.rs
use crate::error::ArcaneError;
use serde_json::{json, Value};

const RECOVERY_SCHEMA: &str = "arcane.control-recovery.v1";

pub fn recover_control_state(
    input: &Value,
    authenticate: &dyn Fn(&Value) -> bool,
    repair: &dyn Fn(&Value, &Value) -> Result<Value, ArcaneError>,
    verify: &dyn Fn(&Value, &Value) -> bool,
) -> Result<Value, ArcaneError> {
    let control_id = input
        .get("controlId")
        .and_then(Value::as_str)
        .filter(|value| !value.is_empty())
        .ok_or_else(|| ArcaneError::typed("ARC_SCHEMA_INVALID", "controlId must be a non-empty string"))?;
    let authorization = input
        .get("authorization")
        .and_then(Value::as_object)
        .ok_or_else(|| ArcaneError::typed("ARC_SCHEMA_INVALID", "authorization must be an object"))?;
    if authorization.get("controlId").and_then(Value::as_str) != Some(control_id)
        || authorization.get("operation").and_then(Value::as_str) != Some("RECOVER_CONTROL_STATE")
    {
        return Ok(json!({
            "schema": RECOVERY_SCHEMA,
            "allowed": false,
            "code": "ARC_RECOVERY_SCOPE_DENIED",
            "controlId": control_id,
        }));
    }
    let authorization_value = Value::Object(authorization.clone());
    if !authenticate(&authorization_value) {
        return Ok(json!({
            "schema": RECOVERY_SCHEMA,
            "allowed": false,
            "code": "ARC_RECOVERY_AUTH_REQUIRED",
            "controlId": control_id,
        }));
    }
    let state = input.get("state").cloned().unwrap_or(Value::Null);
    let context = json!({ "controlId": control_id, "authorization": authorization });
    let recovered = match repair(&state, &context) {
        Ok(value) => value,
        Err(error) => {
            return Ok(json!({
                "schema": RECOVERY_SCHEMA,
                "allowed": false,
                "code": "ARC_RECOVERY_QUARANTINED",
                "controlId": control_id,
                "quarantine": {
                    "originalState": state,
                    "reasonCode": error.code(),
                },
            }));
        }
    };
    if !verify(&recovered, &context) {
        return Ok(json!({
            "schema": RECOVERY_SCHEMA,
            "allowed": false,
            "code": "ARC_RECOVERY_VERIFY_FAILED",
            "controlId": control_id,
            "quarantine": {
                "originalState": state,
                "recoveredState": recovered,
                "reasonCode": "ARC_RECOVERY_VERIFY_FAILED",
            },
        }));
    }
    Ok(json!({
        "schema": RECOVERY_SCHEMA,
        "allowed": true,
        "code": "ARC_RECOVERY_RESUMED",
        "controlId": control_id,
        "replacementState": recovered,
        "preservedState": state,
        "verification": "PASS",
    }))
}
